import click
from loguru import logger
from colorama import init, Fore, Style
from Config import COORDINATOR_DEFAULT_PORT
# Thriftå¯¼å…¥
from src.vector_db import CoordinatorService
from src.vector_db.ttypes import SearchRequest
from thrift.transport import TSocket, TTransport
from thrift.protocol import TBinaryProtocol

# åˆå§‹åŒ–å½©è‰²è¾“å‡º
init(autoreset=True)

@click.group()
@click.option("--coord-addr", default=f"127.0.0.1:{COORDINATOR_DEFAULT_PORT}", help="åè°ƒèŠ‚ç‚¹åœ°å€")
@click.pass_context
def cli(ctx, coord_addr):
    """åˆ†å¸ƒå¼å‘é‡æ•°æ®åº“CLIå·¥å…·"""
    ctx.ensure_object(dict)
    # åˆå§‹åŒ–åè°ƒèŠ‚ç‚¹å®¢æˆ·ç«¯
    host, port = coord_addr.split(":")
    transport = TSocket.TSocket(host, int(port))
    transport = TTransport.TBufferedTransport(transport)
    protocol = TBinaryProtocol.TBinaryProtocol(transport)
    ctx.obj["client"] = CoordinatorService.Client(protocol)
    ctx.obj["transport"] = transport
    ctx.obj["coord_addr"] = coord_addr

# èŠ‚ç‚¹ç®¡ç†å‘½ä»¤
@cli.command()
@click.option("--node-id", required=True, help="èŠ‚ç‚¹ID")
@click.option("--node-addr", required=True, help="èŠ‚ç‚¹åœ°å€ï¼ˆhost:portï¼‰")
@click.pass_context
def register_node(ctx, node_id, node_addr):
    """æ³¨å†Œæ•°æ®èŠ‚ç‚¹"""
    client = ctx.obj["client"]
    transport = ctx.obj["transport"]
    try:
        transport.open()
        resp = client.register_node(node_id, node_addr)
        transport.close()
        if resp.success:
            click.echo(Fore.GREEN + f"âœ… {resp.message}")
        else:
            click.echo(Fore.RED + f"âŒ {resp.message}")
    except Exception as e:
        transport.close()
        click.echo(Fore.RED + f"âŒ æ³¨å†Œå¤±è´¥ï¼š{str(e)}")

@cli.command()
@click.pass_context
def list_nodes(ctx):
    """åˆ—å‡ºæ‰€æœ‰æ•°æ®èŠ‚ç‚¹"""
    client = ctx.obj["client"]
    transport = ctx.obj["transport"]
    try:
        transport.open()
        resp = client.list_nodes()
        transport.close()
        if resp.success:
            click.echo(Fore.BLUE + "\nğŸ“Œ æ•°æ®èŠ‚ç‚¹åˆ—è¡¨ï¼š")
            from prettytable import PrettyTable
            table = PrettyTable()
            table.field_names = ["èŠ‚ç‚¹ID", "åœ°å€"]
            for node_id, addr in resp.vector_data.metadata.items():
                table.add_row([node_id, addr])
            click.echo(table)
        else:
            click.echo(Fore.RED + f"âŒ {resp.message}")
    except Exception as e:
        transport.close()
        click.echo(Fore.RED + f"âŒ è·å–èŠ‚ç‚¹å¤±è´¥ï¼š{str(e)}")

# å‘é‡æ“ä½œå‘½ä»¤
@cli.command()
@click.option("--key", required=True, help="å‘é‡Key")
@click.option("--vector", required=True, help="å‘é‡ï¼ˆé€—å·åˆ†éš”ï¼‰")
@click.option("--metadata", help="å…ƒæ•°æ®ï¼ˆkey=value,key2=value2ï¼‰")
@click.pass_context
def put(ctx, key, vector, metadata):
    """å†™å…¥/æ›´æ–°å‘é‡"""
    # è§£æå‘é‡
    try:
        vector_list = [float(x.strip()) for x in vector.split(",")]
    except:
        click.echo(Fore.RED + "âŒ å‘é‡æ ¼å¼é”™è¯¯ï¼šé€—å·åˆ†éš”çš„æ•°å­—")
        return
    # è§£æå…ƒæ•°æ®
    meta_dict = {}
    if metadata:
        try:
            for item in metadata.split(","):
                k, v = item.split("=")
                meta_dict[k.strip()] = v.strip()
        except:
            click.echo(Fore.RED + "âŒ å…ƒæ•°æ®æ ¼å¼é”™è¯¯ï¼škey=value,key2=value2")
            return

    # æ„é€ è¯·æ±‚
    from src.vector_db.ttypes import VectorData
    data = VectorData(
        key=key,
        vector=vector_list,
        metadata=meta_dict
    )

    # å‘é€è¯·æ±‚
    client = ctx.obj["client"]
    transport = ctx.obj["transport"]
    try:
        transport.open()
        resp = client.put(data)
        transport.close()
        if resp.success:
            click.echo(Fore.GREEN + f"âœ… å†™å…¥æˆåŠŸï¼Key={key}")
        else:
            click.echo(Fore.RED + f"âŒ {resp.message}")
    except Exception as e:
        transport.close()
        click.echo(Fore.RED + f"âŒ ç½‘ç»œé”™è¯¯ï¼š{str(e)}")

@cli.command()
@click.option("--key", required=True, help="å‘é‡Key")
@click.pass_context
def delete(ctx, key):
    """åˆ é™¤å‘é‡"""
    client = ctx.obj["client"]
    transport = ctx.obj["transport"]
    try:
        transport.open()
        resp = client.delete(key)
        transport.close()
        if resp.success:
            click.echo(Fore.GREEN + f"âœ… åˆ é™¤æˆåŠŸï¼Key={key}")
        else:
            click.echo(Fore.RED + f"âŒ {resp.message}")
    except Exception as e:
        transport.close()
        click.echo(Fore.RED + f"âŒ ç½‘ç»œé”™è¯¯ï¼š{str(e)}")

@cli.command()
@click.option("--key", required=True, help="å‘é‡Key")
@click.pass_context
def get(ctx, key):
    """è·å–å‘é‡"""
    client = ctx.obj["client"]
    transport = ctx.obj["transport"]
    try:
        transport.open()
        resp = client.get(key)
        transport.close()
        if resp.success:
            data = resp.vector_data
            click.echo(Fore.GREEN + f"âœ… è·å–æˆåŠŸï¼")
            click.echo(Fore.BLUE + f"ğŸ“Œ Keyï¼š{data.key}")
            click.echo(Fore.BLUE + f"ğŸ“Œ å‘é‡ç»´åº¦ï¼š{len(data.vector)}")
            click.echo(Fore.BLUE + f"ğŸ“Œ å…ƒæ•°æ®ï¼š{data.metadata}")
        else:
            click.echo(Fore.RED + f"âŒ {resp.message}")
    except Exception as e:
        transport.close()
        click.echo(Fore.RED + f"âŒ ç½‘ç»œé”™è¯¯ï¼š{str(e)}")

@cli.command()
@click.option("--query-vec", required=True, help="æŸ¥è¯¢å‘é‡ï¼ˆé€—å·åˆ†éš”ï¼‰")
@click.option("--top-k", default=5, help="è¿”å›Top-K")
@click.option("--filter", help="è¿‡æ»¤æ¡ä»¶ï¼ˆkey=valueï¼‰")
@click.option("--threshold", default=0.0, help="ç›¸ä¼¼åº¦é˜ˆå€¼")
@click.pass_context
def search(ctx, query_vec, top_k, filter, threshold):
    """å‘é‡æ£€ç´¢"""
    # è§£æå‘é‡
    try:
        query_list = [float(x.strip()) for x in query_vec.split(",")]
    except:
        click.echo(Fore.RED + "âŒ å‘é‡æ ¼å¼é”™è¯¯ï¼šé€—å·åˆ†éš”çš„æ•°å­—")
        return
    # è§£æè¿‡æ»¤æ¡ä»¶
    filter_dict = {}
    if filter:
        try:
            for item in filter.split(","):
                k, v = item.split("=")
                filter_dict[k.strip()] = v.strip()
        except:
            click.echo(Fore.RED + "âŒ è¿‡æ»¤æ¡ä»¶æ ¼å¼é”™è¯¯ï¼škey=value,key2=value2")
            return

    # æ„é€ è¯·æ±‚
    req = SearchRequest(
        query_vector=query_list,
        top_k=top_k,
        filter=filter_dict,
        threshold=threshold
    )

    # å‘é€è¯·æ±‚
    client = ctx.obj["client"]
    transport = ctx.obj["transport"]
    try:
        transport.open()
        resp = client.search(req)
        transport.close()
        if resp.success:
            res = resp.search_result
            click.echo(Fore.GREEN + f"âœ… æ£€ç´¢æˆåŠŸï¼å…±{len(res.keys)}æ¡ç»“æœ")
            from prettytable import PrettyTable
            table = PrettyTable()
            table.field_names = ["æ’å", "Key", "ç›¸ä¼¼åº¦åˆ†æ•°", "å…ƒæ•°æ®"]
            for i, (k, s, vec) in enumerate(zip(res.keys, res.scores, res.vectors)):
                table.add_row([i+1, k, f"{s:.4f}", vec.metadata])
            click.echo(table)
        else:
            click.echo(Fore.RED + f"âŒ {resp.message}")
    except Exception as e:
        transport.close()
        click.echo(Fore.RED + f"âŒ ç½‘ç»œé”™è¯¯ï¼š{str(e)}")

if __name__ == "__main__":
    cli()